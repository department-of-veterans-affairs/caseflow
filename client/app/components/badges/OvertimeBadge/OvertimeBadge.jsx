import PropTypes from 'prop-types';
import * as React from 'react';
import { connect } from 'react-redux';

import Badge from '../Badge';
import { COLORS } from 'app/constants/AppConstants';

/**
 * Component to display if the provided appeal has been approved for overtime work.
 */

class OvertimeBadge extends React.PureComponent {
  render = () => {
    const { appeal, canViewOvertimeStatus } = this.props;

    if (!appeal.overtime || !canViewOvertimeStatus) {
      return null;
    }

    const tooltipText = <div>
      This case has been approved to be worked in overtime hours
    </div>;

    return <Badge name="overtime" displayName="OT" color={COLORS.GREY_DARK} tooltipText={tooltipText} id={appeal.id}
      ariaLabel="overtime" />;
  }
}

OvertimeBadge.propTypes = {
  appeal: PropTypes.object,
  canViewOvertimeStatus: PropTypes.bool
};

const mapStateToProps = (state) => ({ canViewOvertimeStatus: state.ui.canViewOvertimeStatus });

export default connect(mapStateToProps)(OvertimeBadge);
