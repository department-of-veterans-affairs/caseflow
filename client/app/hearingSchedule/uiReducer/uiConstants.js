export const ACTIONS = {
  SET_USER_CSS_ID: 'SET_USER_CSS_ID'
};