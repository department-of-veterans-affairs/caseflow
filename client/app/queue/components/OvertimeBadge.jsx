import PropTypes from 'prop-types';
import { css } from 'glamor';
import _ from 'lodash';
import * as React from 'react';
import { connect } from 'react-redux';

import Tooltip from '../../components/Tooltip';
import { COLORS } from '../../constants/AppConstants';

/**
 * This component can accept an appeal
 * e.g.,
 *   <OvertimeBadge appeal={appeal} />
 */

class OvertimeBadge extends React.PureComponent {
  render = () => {
    const { appeal, canViewOvertimeStatus } = this.props;

    if (!appeal.overtime || !canViewOvertimeStatus) {
      return null;
    }

    // TODO: Throw into copy and talk to geronimo
    const tooltipText = <div>
      This case has been approved to be worked in overtime hours
    </div>;

    const badgeStyling = css({
      ...this.props.badgeStyling,
      background: COLORS.GREY_DARK
    });

    return <div className="cf-overtime-badge">
      <Tooltip id={`badge-${appeal.id}`} text={tooltipText} position="bottom">
        <span {...badgeStyling}>OT</span>
      </Tooltip>
    </div>;
  }
}

OvertimeBadge.propTypes = {
  appeal: PropTypes.object,
  badgeStyling: PropTypes.object
};

const mapStateToProps = (state) => ({ canViewOvertimeStatus: state.ui.canViewOvertimeStatus });

export default connect(mapStateToProps)(OvertimeBadge);
