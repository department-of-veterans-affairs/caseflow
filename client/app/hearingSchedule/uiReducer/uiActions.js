import { ACTIONS } from './uiConstants';

export const setUserCssId = (cssId) => ({
  type: ACTIONS.SET_USER_CSS_ID,
  payload: {
    cssId
  }
});
