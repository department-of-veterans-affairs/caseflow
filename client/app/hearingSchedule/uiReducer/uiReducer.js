import { update } from '../../util/ReducerUtil';
import { ACTIONS } from './uiConstants';

const initialState = {
  userCssId: ''
};

const hearingScheduleUiReducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case ACTIONS.SET_USER_CSS_ID:
      return update(state, {
        userCssId: { $set: action.payload.cssId }
      });
    default:
      return state;
  }
};

export default hearingScheduleUiReducer;
