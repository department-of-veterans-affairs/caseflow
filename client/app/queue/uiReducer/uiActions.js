/* eslint-disable no-console */
import { ACTIONS } from './uiConstants';
import ApiUtil from '../../util/ApiUtil';

export const resetErrorMessages = () => ({
  type: ACTIONS.RESET_ERROR_MESSAGES
});

export const setCanEditAod = (canEditAod) => ({
  type: ACTIONS.SET_CAN_EDIT_AOD,
  payload: {
    canEditAod
  }
});

export const showErrorMessage = (errorMessage) => ({
  type: ACTIONS.SHOW_ERROR_MESSAGE,
  payload: {
    errorMessage
  }
});

export const hideErrorMessage = () => ({
  type: ACTIONS.HIDE_ERROR_MESSAGE
});

export const resetSuccessMessages = () => ({
  type: ACTIONS.RESET_SUCCESS_MESSAGES
});

export const showSuccessMessage = (message) => ({
  type: ACTIONS.SHOW_SUCCESS_MESSAGE,
  payload: {
    message
  }
});

export const hideSuccessMessage = () => ({
  type: ACTIONS.HIDE_SUCCESS_MESSAGE
});

export const highlightInvalidFormItems = (highlight) => ({
  type: ACTIONS.HIGHLIGHT_INVALID_FORM_ITEMS,
  payload: {
    highlight
  }
});

export const setSelectingJudge = (selectingJudge) => ({
  type: ACTIONS.SET_SELECTING_JUDGE,
  payload: {
    selectingJudge
  }
});

const saveSuccess = (message, response) => (dispatch) => {
  dispatch(showSuccessMessage(message));
  dispatch({ type: ACTIONS.SAVE_SUCCESS });

  return Promise.resolve(response);
};

const saveFailure = (err) => (dispatch) => {
  const { response } = err;
  let uiErrorMessage;

  try {
    uiErrorMessage = response.body;
  } catch (ex) {
    // the default case if there is no `text` node in the response (ie the backend did not return sufficient info)
    uiErrorMessage = {
      errors: [{
        title: 'Error',
        detail: 'There was an error processing your request. ' +
        'Please retry your action and contact support if errors persist.'
      }]
    };
  }

  dispatch(showErrorMessage(uiErrorMessage.errors[0]));
  dispatch({ type: ACTIONS.SAVE_FAILURE });
  // the promise rejection below is also uncaught
  // but this seems to be by design since that's the same as the frontend handling and throwing an error

  return Promise.reject(new Error(response.text));
};

export const requestSave = (
  url, params, successMessage, verb = 'post'
) => (dispatch) => {
  dispatch(hideErrorMessage());
  dispatch(hideSuccessMessage());

  dispatch({ type: ACTIONS.REQUEST_SAVE });

  return ApiUtil[verb](url, params).
    then(
      (resp) => dispatch(saveSuccess(successMessage, resp)),

    ).
    catch((err) => dispatch(saveFailure(err)));
};

export const requestGet = (url, params, successMessage) =>
  requestSave(url, params, successMessage, 'get');
export const requestPatch = (url, params, successMessage) =>
  requestSave(url, params, successMessage, 'patch');
export const requestUpdate = (url, params, successMessage) =>
  requestSave(url, params, successMessage, 'put');
export const requestDelete = (url, params, successMessage) =>
  requestSave(url, params, successMessage, 'delete');

export const setSavePending = () => ({
  type: ACTIONS.REQUEST_SAVE
});

export const resetSaveState = () => ({
  type: ACTIONS.RESET_SAVE_STATE
});

export const showModal = (modalType) => ({
  type: ACTIONS.SHOW_MODAL,
  payload: { modalType }
});

export const hideModal = (modalType) => ({
  type: ACTIONS.HIDE_MODAL,
  payload: { modalType }
});

export const setFeatureToggles = (featureToggles) => ({
  type: ACTIONS.SET_FEATURE_TOGGLES,
  payload: { featureToggles }
});

export const setUserRole = (userRole) => ({
  type: ACTIONS.SET_USER_ROLE,
  payload: { userRole }
});

export const setUserCssId = (cssId) => ({
  type: ACTIONS.SET_USER_CSS_ID,
  payload: { cssId }
});

export const setOrganizations = (organizations) => ({
  type: ACTIONS.SET_ORGANIZATIONS,
  payload: { organizations }
});

export const setActiveOrganization = (id, name, isVso) => ({
  type: ACTIONS.SET_ACTIVE_ORGANIZATION,
  payload: {
    id,
    name,
    isVso
  }
});

export const setUserId = (userId) => ({
  type: ACTIONS.SET_USER_ID,
  payload: { userId }
});

export const setTargetUserCssId = (targetUserCssId) => ({
  type: ACTIONS.SET_TARGET_USER_CSS_ID,
  payload: { targetUserCssId }
});

export const setUserIsVsoEmployee = (userIsVsoEmployee) => ({
  type: ACTIONS.SET_USER_IS_VSO_EMPLOYEE,
  payload: { userIsVsoEmployee }
});

export const setFeedbackUrl = (feedbackUrl) => ({
  type: ACTIONS.SET_FEEDBACK_URL,
  payload: { feedbackUrl }
});

export const setSelectedAssignee = ({ assigneeId }) => ({
  type: ACTIONS.SET_SELECTED_ASSIGNEE,
  payload: {
    assigneeId
  }
});

export const setSelectedAssigneeSecondary = ({ assigneeId }) => ({
  type: ACTIONS.SET_SELECTED_ASSIGNEE_SECONDARY,
  payload: {
    assigneeId
  }
});

export const toggleVeteranCaseList = () => ({
  type: ACTIONS.TOGGLE_VETERAN_CASE_LIST
});

export const showVeteranCaseList = () => ({
  type: ACTIONS.SHOW_VETERAN_CASE_LIST
});

export const hideVeteranCaseList = () => ({
  type: ACTIONS.HIDE_VETERAN_CASE_LIST
});

export const setHearingDay = (hearingDay) => ({
  type: ACTIONS.SET_HEARING_DAY,
  payload: hearingDay
});
