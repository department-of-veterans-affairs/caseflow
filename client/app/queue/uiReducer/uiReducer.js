import { update } from '../../util/ReducerUtil';
import { ACTIONS } from './uiConstants';

const initialSaveState = {
  savePending: false,
  saveSuccessful: null
};

export const initialState = {
  selectingJudge: false,
  highlightFormItems: false,
  messages: {
    success: null,
    error: null
  },
  saveState: initialSaveState,
  modals: {},
  userRole: '',
  userCssId: '',
  organizations: [],
  activeOrganization: {
    id: null,
    name: null,
    isVso: false
  },
  userIsVsoEmployee: false,
  feedbackUrl: '#',
  loadedUserId: null,
  selectedAssignee: null,
  selectedAssigneeSecondary: null,
  veteranCaseListIsVisible: false,
  canEditAod: false,
  hearingDay: {
    hearingDate: null,
    regionalOffice: null
  }
};

const setMessageState = (state, message, msgType) => update(state, {
  messages: {
    [msgType]: {
      $set: message
    }
  }
});

const setErrorMessageState = (state, message) => setMessageState(state, message, 'error');
const hideErrorMessage = (state) => setErrorMessageState(state, null);
const showErrorMessage = (state, errorMsg = { title: 'Error' }) => setErrorMessageState(state, errorMsg);

const setSuccessMessageState = (state, message) => setMessageState(state, message, 'success');
const hideSuccessMessage = (state) => setSuccessMessageState(state, null);
const showSuccessMessage = (state, message = { title: 'Success' }) => setSuccessMessageState(state, message);

const setModalState = (state, visibility, modalType) => update(state, {
  modals: {
    [modalType]: {
      $set: visibility
    }
  }
});

const showModal = (state, modalType) => setModalState(state, true, modalType);
const hideModal = (state, modalType) => setModalState(state, false, modalType);

const workQueueUiReducer = (state = initialState, action = {}) => {
  switch (action.type) {
  case ACTIONS.SET_SELECTING_JUDGE:
    return update(state, {
      selectingJudge: { $set: action.payload.selectingJudge }
    });
  case ACTIONS.SET_CAN_EDIT_AOD:
    return update(state, {
      canEditAod: { $set: action.payload.canEditAod }
    });
  case ACTIONS.HIGHLIGHT_INVALID_FORM_ITEMS:
    return update(state, {
      highlightFormItems: {
        $set: action.payload.highlight
      }
    });
  case ACTIONS.REQUEST_SAVE:
    return update(state, {
      saveState: {
        savePending: { $set: true },
        saveSuccessful: { $set: null }
      }
    });
  case ACTIONS.SAVE_SUCCESS:
    return update(state, {
      saveState: {
        savePending: { $set: false },
        saveSuccessful: { $set: true }
      }
    });
  case ACTIONS.SAVE_FAILURE:
    return update(state, {
      saveState: {
        savePending: { $set: false },
        saveSuccessful: { $set: false }
      }
    });
  case ACTIONS.RESET_SAVE_STATE:
    return update(state, {
      saveState: { $set: initialSaveState }
    });
  case ACTIONS.RESET_ERROR_MESSAGES:
    return update(state, {
      messages: {
        error: { $set: null }
      }
    });
  case ACTIONS.RESET_SUCCESS_MESSAGES:
    return update(state, {
      messages: {
        success: { $set: null }
      }
    });
  case ACTIONS.HIDE_ERROR_MESSAGE:
    return hideErrorMessage(state);
  case ACTIONS.SHOW_ERROR_MESSAGE:
    return showErrorMessage(state, action.payload.errorMessage);
  case ACTIONS.SHOW_SUCCESS_MESSAGE:
    return showSuccessMessage(state, action.payload.message);
  case ACTIONS.HIDE_SUCCESS_MESSAGE:
    return hideSuccessMessage(state);
  case ACTIONS.SHOW_MODAL:
    return showModal(state, action.payload.modalType);
  case ACTIONS.HIDE_MODAL:
    return hideModal(state, action.payload.modalType);
  case ACTIONS.TOGGLE_VETERAN_CASE_LIST:
    return update(state, {
      veteranCaseListIsVisible: { $set: !state.veteranCaseListIsVisible }
    });
  case ACTIONS.SHOW_VETERAN_CASE_LIST:
    return update(state, {
      veteranCaseListIsVisible: { $set: true }
    });
  case ACTIONS.HIDE_VETERAN_CASE_LIST:
    return update(state, {
      veteranCaseListIsVisible: { $set: false }
    });
  case ACTIONS.SET_USER_ID:
    return update(state, {
      loadedUserId: { $set: action.payload.userId }
    });
  case ACTIONS.SET_USER_ROLE:
    return update(state, {
      userRole: { $set: action.payload.userRole }
    });
  case ACTIONS.SET_USER_CSS_ID:
    return update(state, {
      userCssId: { $set: action.payload.cssId }
    });
  case ACTIONS.SET_TARGET_USER_CSS_ID:
    return update(state, {
      targetUserCssId: { $set: action.payload.targetUserCssId }
    });
  case ACTIONS.SET_USER_IS_VSO_EMPLOYEE:
    return update(state, {
      userIsVsoEmployee: { $set: action.payload.userIsVsoEmployee }
    });
  case ACTIONS.SET_FEEDBACK_URL:
    return update(state, {
      feedbackUrl: { $set: action.payload.feedbackUrl }
    });
  case ACTIONS.SET_SELECTED_ASSIGNEE:
    return update(state, {
      selectedAssignee: {
        $set: action.payload.assigneeId
      }
    });
  case ACTIONS.SET_SELECTED_ASSIGNEE_SECONDARY:
    return update(state, {
      selectedAssigneeSecondary: {
        $set: action.payload.assigneeId
      }
    });
  case ACTIONS.SET_ORGANIZATIONS:
    return update(state, {
      organizations: {
        $set: action.payload.organizations
      }
    });
  case ACTIONS.SET_ACTIVE_ORGANIZATION:
    return update(state, {
      activeOrganization: {
        id: { $set: action.payload.id },
        name: { $set: action.payload.name },
        isVso: { $set: action.payload.isVso }
      }
    });
  case ACTIONS.SET_HEARING_DAY:
    return update(state, {
      hearingDay: {
        $set: action.payload
      }
    });
  default:
    return state;
  }
};

export default workQueueUiReducer;
