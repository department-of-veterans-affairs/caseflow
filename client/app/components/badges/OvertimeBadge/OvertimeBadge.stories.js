import React from 'react';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';

import OvertimeBadge from './OvertimeBadge';
import rootReducer from 'app/queue/reducers';
import { setCanViewOvertimeStatus } from 'app/queue/uiReducer/uiActions';

export default {
  title: 'Commons/Components/Badges/OT Badge',
  component: OvertimeBadge,
  parameters: {
    layout: 'centered',
  },
  args: {
    appeal: {
      overtime: true
    },
    canViewOvertimeStatus: true,
    featureToggles: {
      overtime_revamp: true
    },
  }
};

const getStore = () => createStore(rootReducer, applyMiddleware(thunk));
const store = getStore();
store.dispatch(setCanViewOvertimeStatus(true));

const Template = (args) => (
  <Provider store={store}>
    <OvertimeBadge {...args} />
  </Provider>
);

export const OTBadge = Template.bind({});
