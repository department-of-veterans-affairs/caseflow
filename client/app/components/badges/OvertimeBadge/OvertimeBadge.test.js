import React from 'react';
import { axe } from 'jest-axe';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import { render, screen, waitFor } from '@testing-library/react';

import OvertimeBadge from './OvertimeBadge';
import rootReducer from 'app/queue/reducers';
import { setCanViewOvertimeStatus } from 'app/queue/uiReducer/uiActions';

describe('OvertimeBadge', () => {

  const tooltipText = 'OT';

  const getStore = () => createStore(rootReducer, applyMiddleware(thunk));

  const setupOvertimeBadge = (store, appeal) => render(
    <Provider store={store}>
      <OvertimeBadge appeal={appeal} />
    </Provider>
  );

  it('renders correctly', () => {
    const store = getStore();

    store.dispatch(setCanViewOvertimeStatus(true));
    const { container } = setupOvertimeBadge(store, { overtime: true });

    expect(container).toMatchSnapshot();
  });

  it('doesn\'t show if the appeal is not marked as overtime', () => {
    const store = getStore();

    store.dispatch(setCanViewOvertimeStatus(true));
    const component = setupOvertimeBadge(store, { overtime: false });

    expect(component.queryByRole('tooltip')).toBeNull;
  });

  it('doesn\'t show if the user cannot view overtime badges', () => {
    const store = getStore();

    store.dispatch(setCanViewOvertimeStatus(false));
    const component = setupOvertimeBadge(store, { overtime: true });

    expect(component.queryByRole('tooltip')).toBeNull;
  });

  it('does show if the appeal is marked as overtime and the user can view overtime badges', async () => {
    const store = getStore();

    store.dispatch(setCanViewOvertimeStatus(true));
    setupOvertimeBadge(store, { overtime: true });

    await waitFor(() => {
      expect(screen.getByText(tooltipText)).toBeInTheDocument();
    });
  });

  it('passes a11y', async () => {
    const store = getStore();

    store.dispatch(setCanViewOvertimeStatus(true));
    const { container } = setupOvertimeBadge(store, { overtime: true });

    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });
});
