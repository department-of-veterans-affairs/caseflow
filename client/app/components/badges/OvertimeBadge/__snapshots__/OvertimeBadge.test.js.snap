// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OvertimeBadge renders correctly 1`] = `
<div>
  <div
    class="cf-overtime-badge"
  >
    <span
      aria-describedby="badge-undefined"
      aria-label="overtime"
      currentitem="false"
      data-css-13kydni=""
      data-event="focus mouseenter"
      data-event-off="mouseleave keydown"
      data-for="badge-undefined"
      data-tip="true"
      tabindex="0"
    >
      OT
    </span>
    <span
      data-css-12sli69=""
    >
      <div
        class="__react_component_tooltip t00000000-0000-0000-0000-000000000000 place-bottom type-dark"
        data-id="tooltip"
        id="badge-undefined"
        role="tooltip"
      >
        <style>
          
  	.t00000000-0000-0000-0000-000000000000 {
	    color: #fff;
	    background: #222;
	    border: 1px solid transparent;
  	}

  	.t00000000-0000-0000-0000-000000000000.place-top {
        margin-top: -10px;
    }
    .t00000000-0000-0000-0000-000000000000.place-top::before {
        border-top: 8px solid transparent;
    }
    .t00000000-0000-0000-0000-000000000000.place-top::after {
        border-left: 8px solid transparent;
        border-right: 8px solid transparent;
        bottom: -6px;
        left: 50%;
        margin-left: -8px;
        border-top-color: #222;
        border-top-style: solid;
        border-top-width: 6px;
    }

    .t00000000-0000-0000-0000-000000000000.place-bottom {
        margin-top: 10px;
    }
    .t00000000-0000-0000-0000-000000000000.place-bottom::before {
        border-bottom: 8px solid transparent;
    }
    .t00000000-0000-0000-0000-000000000000.place-bottom::after {
        border-left: 8px solid transparent;
        border-right: 8px solid transparent;
        top: -6px;
        left: 50%;
        margin-left: -8px;
        border-bottom-color: #222;
        border-bottom-style: solid;
        border-bottom-width: 6px;
    }

    .t00000000-0000-0000-0000-000000000000.place-left {
        margin-left: -10px;
    }
    .t00000000-0000-0000-0000-000000000000.place-left::before {
        border-left: 8px solid transparent;
    }
    .t00000000-0000-0000-0000-000000000000.place-left::after {
        border-top: 5px solid transparent;
        border-bottom: 5px solid transparent;
        right: -6px;
        top: 50%;
        margin-top: -4px;
        border-left-color: #222;
        border-left-style: solid;
        border-left-width: 6px;
    }

    .t00000000-0000-0000-0000-000000000000.place-right {
        margin-left: 10px;
    }
    .t00000000-0000-0000-0000-000000000000.place-right::before {
        border-right: 8px solid transparent;
    }
    .t00000000-0000-0000-0000-000000000000.place-right::after {
        border-top: 5px solid transparent;
        border-bottom: 5px solid transparent;
        left: -6px;
        top: 50%;
        margin-top: -4px;
        border-right-color: #222;
        border-right-style: solid;
        border-right-width: 6px;
    }
  
        </style>
        <div>
          This case has been approved to be worked in overtime hours
        </div>
      </div>
    </span>
  </div>
</div>
`;
